### PR TITLE
[Driver][SYCL][clang-linker-wrapper] Add support for additional linker wrapper options

### PR DESCRIPTION
**Pull request by mdtoguchi**
Adds support for a handful of linker wrapper options that are part of the new offload model support.
  --sycl-device-libraries=libs provides a comma separated list of
    libraries that are used during the device link
  --sycl-device-library-location=dir the location in which the device
    libraries reside.
  --sycl-post-link-options=opts all options as specified by the user
    via the -Xdevice-post-link option.
  --llvm-spirv-options=opts all options as specified by the user via
    the -Xspirv-translator option.

Does some refactoring for what libraries are added, moving the functionality to the SYCL based toolchain.
These options are documented in the following design document: sycl/doc/design/OffloadDesign.md

### CODE REVIEW

**Comment by asudarsa** [comment]
> Adds support for a handful of linker wrapper options that are part of the new offload model support. --sycl-device-libraries=libs provides a comma separated list of libraries that are used during the device link --sycl-device-library-location=dir the location in which the device libraries reside. --sycl-post-link-options=opts all options as specified by the user via the -Xdevice-post-link option. --llvm-spirv-options=opts all options as specified by the user via the -Xspirv-translator option.
> 
> Does some refactoring for what libraries are added, moving the functionality to the SYCL based toolchain.

Hi @mdtoguchi 

Thanks for the PR.
In the description, Please add:
These options are documented in the following design document: sycl/doc/design/OffloadDesign.md

Also, for sycl-post-link and spirv-llvm-translator, the clang-linker-wrapper currently expects ALL arguments to be passed in from the driver, not just the ones passed in by the user. For example, the '-split=<split type>' is expected to passed in via --sycl-post-link-options.

Thanks

**Comment by mdtoguchi** [comment]
> > Adds support for a handful of linker wrapper options that are part of the new offload model support. --sycl-device-libraries=libs provides a comma separated list of libraries that are used during the device link --sycl-device-library-location=dir the location in which the device libraries reside. --sycl-post-link-options=opts all options as specified by the user via the -Xdevice-post-link option. --llvm-spirv-options=opts all options as specified by the user via the -Xspirv-translator option.
> > Does some refactoring for what libraries are added, moving the functionality to the SYCL based toolchain.
> 
> Hi @mdtoguchi
> 
> Thanks for the PR. In the description, Please add: These options are documented in the following design document: sycl/doc/design/OffloadDesign.md
> 
> Also, for sycl-post-link and spirv-llvm-translator, the clang-linker-wrapper currently expects ALL arguments to be passed in from the driver, not just the ones passed in by the user. For example, the '-split=' is expected to passed in via --sycl-post-link-options.
> 
> Thanks

@asudarsa, I'll update the description to add a pointer to the design.  As for the additional options for the translator and the post-link step, I think that will need to be done in a separate PR.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -134,6 +136,173 @@ bool SYCL::shouldDoPerObjectFileLinking(const Compilation &C) {
...
+      if (SYCLTriple == Triple) {
+        const toolchains::SYCLToolChain *SYCLTC =
+            static_cast<const toolchains::SYCLToolChain *>(TI->second);
+        SYCLTC->TranslateBackendTargetArgs(Triple, Args, TargArgs);
+        break;
+      }
+    }
+
+    auto checkBF = [=](std::string &Dev) {
+      static const std::regex BFFs("pvc.*|ats.*");
```
Comment (maksimsab) [inline clang/lib/Driver/ToolChains/SYCL.cpp:168]: nit: We can use[`llvm::StringRef::starts_with()`](https://llvm.org/doxygen/classllvm_1_1StringRef.html#a2cd8968ff703aaeb395dcd63f6805ff1) function for this until C++20 turns on.

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: I have updated to use `starts_with`

**Comment by mdtoguchi** [comment]
@intel/dpcpp-clang-driver-reviewers, please take a look.

**Review by hchilama — APPROVED** [review]

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, this looks ready to go - thanks!

**Pull request merged**